doc/API.md

```diff
@@ -608,6 +613,120 @@ In axe-core v1 the main method for axe was `axe.a11yCheck()`. This method was re
 - .a11yCheck requires a context object, and so will not fall back to the document root.
 - .a11yCheck does not return a Promise.
 
+### Virtual DOM Utilities
+
+#### API Name: axe.utils.getFlattenedTree
+
+##### Description
+
+Recursvely return an array containing the virtual DOM tree for the node specified, excluding comment nodes and shadow DOM nodes `&lt;content> and `&lt;slot>`. This method will return a composed tree containing both light and shadow DOM, if applicable.
+
+##### Synopsis
+
+```javascript
+var element = document.body;
+axe.utils.getFlattenedTree(element, shadowId)
+```
+
+##### Parameters
+ - `node` – node. The current HTML node for which you want a flattened DOM tree.
+ - `shadowId` – string(optional). ID of the shadow DOM that is the closest shadow ancestor of the node
+
+##### Returns
+
+An array of objects, where each object is a virtualNode:
+
+```javascript
+[{
+  actualNode: body,
+  children: [virtualNodes],
+  shadowId: undefined
+}]
+```
+
+#### API Name: axe.utils.getNodeFromTree
+
+##### Description
+
+Recursively return a single node from a virtual DOM tree. This is commonly used in rules and checks where the node is readily available without querying the DOM.
+
+##### Synopsis
+
+```javascript
+axe.utils.getNodeFromTree(axe._tree[0], node);
+```
+
+##### Parameters
+
+  - `vNode` – object. The flattened DOM tree to fetch a virtual node from
+  - `node` – node. The HTML DOM node for which you need a virtual representation
+
+##### Returns
+
+An object containing the virtualNode:
+
+```javascript
+{
+  actualNode: div,
+  children: [virtualNodes],
+  shadowId: undefined
+}
+```
+
+#### API Name: axe.utils.querySelectorAll
+
+##### Description
+
+A querySelectorAll implementation that works on the virtual DOM and Shadow DOM by manually walking the tree instead of relying on DOM API methods which don't step into Shadow DOM.
```

Comment:
I mean this implementation walks the flattened tree. It is like `document|Element.querySelectorAll` except for the flattened tree rather than the DOM (light DOM).